src/mysql_conn.erl

```diff
@@ -91,54 +92,109 @@ init(Opts) ->
         N when N > 0 -> N
     end,
 
+    State0 = #state{
+        tcp_opts = TcpOpts,
+        ssl_opts = SSLOpts,
+        host = Host, port = Port,
+        user = User, password = Password,
+        database = Database,
+        queries = Queries, prepares = Prepares,
+        log_warnings = LogWarn,
+        connect_timeout = ConnectTimeout,
+        ping_timeout = PingTimeout,
+        query_timeout = QueryTimeout,
+        query_cache_time = QueryCacheTime,
+        cap_found_rows = (SetFoundRows =:= true)
+    },
+
+    case proplists:get_value(connect_mode, Opts, synchronous) of
+        synchronous ->
+            case do_connect(State0) of
+                {ok, State1} ->
+                    {ok, State1};
+                {error, Reason} ->
+                    {stop, Reason}
+            end;
+        asynchronous ->
+            gen_server:cast(self(), connect),
+            {ok, State0};
+        lazy ->
+            {ok, State0}
+    end.
+
+do_connect(#state{connect_timeout = ConnectTimeout} = State) ->
+    Self = self(),
+    Pid = spawn_link(
+        fun () ->
+            case do_connect1(State) of
+                {ok, #state{sockmod = SockMod, socket = Socket} = State1} ->
```

Comment:
Is this spawned just to be able to set a timeout using receive after? It would be more clean connect in the same process and just use recv timeouts on tcp/ssl. The timeout value doesn't need to be a hard limit IMO.

Reply:
Yes, just for that reason.

I had built into connect etc in the beginning, but found that I had to push that timeout around very often: `gen_tcp:connect`, `ssl:connect` (where it is currently hardcoded to `5000`), and also to all the `send` and `recv` calls that are involved in the various steps of the handshake to be consistent. And actually, I'd have to keep track and reduce it with every step taken to stay close to the overall timeout, otherwise the overall time taken might add up and get quite far from the desired timeout value. Also, there needs to be timeout handling put in place like everywhere, eg without a timeout, `gen_tcp:connect` will always return `{ok, Socket}` sooner or later (unless there are other errors, of course), while with a timeout it may also return `{error, timeout}` or something, etc, same for all the `send` and `recv` calls.

For that reason, I switched to connecting in a separate process, which can run much as it does now, but gets killed from the outside if it takes too long for the overall process.

Reply:
Fine then!

So I guess the ownership of the socket gets transferred when the connecting process is done? I mean, for `{active, once}` later, messages like `tcp_closed` have to be delivered to the connection process, not to the connecting helper-process which isn't alive anymore.